SJRK-321: added BRANCHES.md and updated README with links to other docs

BlueSlug opened the pull request:
This pull request coincides with another pull request to update the CI job for the staging site to run from `master` rather than `stories-floe-dev`: https://github.com/inclusive-design/ci-service/pull/37

BlueSlug committed:
SJRK-321: added BRANCHES.md and updated README with links to other docs

BlueSlug committed:
SJRK-321: updated documentation to satisfy the Laser Eye Checklist

BlueSlug committed:
SJRK-321: updated test file comments for the Laser Eye Checklist
Also did a bit of light refactoring to make function names more 
consistent (test vs. verify)

BlueSlug commented:
Ready for review

cindyli commented on docs/BRANCHES.md:
```diff
@@ -0,0 +1,16 @@
+| ----------------------------------- | ---- | ----------- |
+| `master`                            | [The Storytelling Project staging site](https://staging-stories.floeproject.org/) | Stories on this site are cleared every 24 hours, it is intended purely as a sandbox demonstration of the Storytelling Tool |
+| `stories-cities-production`         | [Co-designing Inclusive Cities Stories site](https://stories.cities.inclusivedesign.ca/) | Part of the larger [Co-designing Inclusive Cities](https://cities.inclusivedesign.ca/) project |
+| `stories-floe-production`           | [The Storytelling Project production site](https://stories.floeproject.org/) | This is the  |
```
The description is incomplete.

BlueSlug replied:
So it is :) I'll address that presently

cindyli commented on docs/BRANCHES.md:
```diff
@@ -0,0 +1,16 @@
+
+Authoring new stories is normally disabled on all deployed versions except the Staging site.
+
+| Branch name                         | Site | Description |
```
There's another branch `learning-to-learn` that is not listed here. Can you add it in? Thanks.

BlueSlug replied:
I was thinking we should remove this branch and left a comment to that effect on [the Jira page](https://issues.fluidproject.org/browse/SJRK-321), so I didn't mention it, `stories-floe-dev` or `re-review`.

cindyli commented on src/server/middleware/saveStoryWithBinaries.js:
```diff
@@ -41,7 +42,10 @@ fluid.defaults("sjrk.storyTelling.server.middleware.saveStoryWithBinaries", {
 });
 
-// Renames any uploaded files to a pattern of uuid + extension
+/* Renames any uploaded files to a pattern of uuid + extension
```
GPII and infusion projects are using [JSDoc|https://jsdoc.app/] style of comments. These comments can be scanned by JSDoc tools to generate static documentation. You probably should consider to use the same pattern.

BlueSlug replied:
I thought about doing that as well, but figured that might be more appropriately done in its own Jira. It's an interesting question though, and I did make an effort to make the comments on all functions quasi-JSDoc like and consistent with each other so that the work to convert them wouldn't be too difficult.

Do you think we should take this on as part of this Jira or leave it for later?

cindyli replied:
If we decide to use JSDoc style, it might be less effort from now on to apply it to any new or changed comments in each pull request. It would be a lots of effort to add comments for all functions, but for comments you are adding/modifying in this pull request, it would be a good start to apply JSDoc in it.

BlueSlug replied:
Oh, I did the work to add comments for all functions as part of this PR :) so it's more just a matter of converting the style from my custom one to JSDoc style

BlueSlug committed:
SJRK-321: filling in missing description

BlueSlug committed:
SJRK-321: converted function comments to JSDoc format
Also did some light refactoring while I was at it

BlueSlug committed:
Merge remote-tracking branch 'upstream/master' into SJRK-321

BlueSlug commented:
I noticed the server test files are also missing comments, so I'll add those as well

BlueSlug committed:
SJRK-321: added missing server comments, renamed functions for accuracy
Also fixed a minor issue in singleNodeConfiguration.js where the usersDb 
and replicatorDb names were swapped

BlueSlug commented:
Ready for review

cindyli commented on src/server/db/story-dbConfiguration.js:
```diff
@@ -115,8 +126,18 @@ sjrk.storyTelling.server.storiesDb.storiesByIdFunction = function (doc) {
+* @param {Object} userCtx - the user and their roles
+* @param {Object} secObj - the Security Object
+*/
 sjrk.storyTelling.server.storiesDb.validateFunction = function (newDoc, oldDoc, userCtx, secObj) {
```
The last 2 arguments are defined but never used in the function. Shall they be removed?

cindyli replied:
I would be helpful to improve the linting to catch this type of issue.

BlueSlug replied:
I believe I left them in thinking the database would complain if they weren't present, but I've just tested the function with only the first arg present (only `newDoc` is referred to in the function body) and everything seems to be working properly. The linter did in fact catch this back when I originally added the function, but since I believed those args to be necessary, I added those arg names to the global eslint definition (line 14). I've now removed them.

That should make things a little easier to document :)

cindyli commented on src/server/middleware/staticMiddlewareSubdirectoryFilter.js:
```diff
@@ -27,6 +28,15 @@ fluid.defaults("sjrk.storyTelling.server.staticMiddlewareSubdirectoryFilter", {
+ * @param {Objec} request - the kettle request
+ * @param {Object} allowedSubdirectories - the list of subdirectories from which to allow serving
+ *
+ * @return {Promise} - a fluid-flavoured promise that returns empty on resolve
```
This is a generic comment for functional docs: please describe the error handling when applicable.

BlueSlug replied:
I found [this comment](https://github.com/jsdoc/jsdoc/issues/1197#issuecomment-312948746) by the main developer of JSDoc, which suggests that a promise be documented only in terms of what is returned on resolution, not rejection. It would seem at some point in the future, the `@reject` syntax will be available, though this comment is nearly 3 years old.

I could use the `@throws` tag, but this may be misleading since the function (in this case anyway) doesn't actually throw any exceptions. Perhaps I could add more detail in the `@return` portion of the documentation, despite the author's suggestion not to. Would that be satisfactory in the short term?

cindyli replied:
Adding more detail in the @return portion of the documentation is a good idea.

cindyli commented on src/server/requestHandlers.js:
```diff
@@ -41,9 +49,16 @@ sjrk.storyTelling.server.handleBrowseStories = function (request, viewDatasource
+ *
+ * @param {Object} response - the raw Browse response
+ *
+ * @return {Object} - a collection of stories as well as the total number and page offset
```
Use @typedef to describe the object structure for important parameters or returns. If the object structure is too dynamic to use @typedef, an example of the object would be helpful too. Thanks.

cindyli commented on src/server/requestHandlers.js:
```diff
@@ -206,7 +252,15 @@ sjrk.storyTelling.server.buildBinaryRenameMap = function (blocks, files) {
+ * @param {Object} file - an image file to process
+ * @param {Object} options - additional options to pass into the rotate call
+ *
+ * @return {Promise} - a fluid-flavoured promise
```
Be more specific with what the returned promise carries. This comment applies to other docs too.

cindyli commented on src/server/requestHandlers.js:
```diff
@@ -304,6 +370,19 @@ sjrk.storyTelling.server.handleDeleteStory = function (request) {
+ * @param {Object} deleteStoryDataSource - a Kettle CouchDB DataSource for deleting a single story
+ * @param {Object} getStoryDataSource - a Kettle CouchDB DataSource for retrieving a single story
+ *
+ * @return {Promise} - a fluid-flavoured promise
```
same as above.

cindyli commented on src/ui/base-page-storyEdit.js:
```diff
@@ -222,25 +227,45 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
-     * - "blockContentValuesForType": an array of values for the block's type
+     *
+     * @param {Object} block - a single story block (sjrk.storyTelling.block)
+     * @param {Object} blockContentValuesForType - an array of values for the block's type
```
The type is `{Object}` while the description says "an array". Please double check and fix. This description sentence is not understandable. Please fix too. Thanks.

BlueSlug replied:
It's a pretty confusing sentence, I agree. I've had a lot of trouble naming this collection succinctly! I'll come up with something :)

BlueSlug replied:
As for the type, JSDoc [recommends](https://jsdoc.app/tags-type.html) either `{Array.<String>}` or `{String[]}`. I'll use the latter if that's alright

cindyli replied:
Sure.

jobara replied:
We've also been using `{String[]}` in other projects lately.

cindyli commented on src/server/requestHandlers.js:
```diff
@@ -159,8 +191,15 @@ sjrk.storyTelling.server.handleSaveStoryWithBinaries = function (request, dataSo
+/**
+ * Persist the story model to couch, with the updated references to where the binaries are saved
+ *
+ * @param {Object} dataSource - a Kettle CouchDB DataSource for saving stories
```
The type should be `{Component}` rather than `{Object}`. Describe what component instance is expected too. For example, "an instance of kettle.dataSource.CouchDB". 

This comment applies thru the entire pull request. Thanks.

cindyli commented on src/ui/base-page-storyEdit.js:
```diff
@@ -222,25 +227,45 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+     * Depending on whether story authoring is enabled, will set/add CSS classes
+     * and HTML properties to show or hide parts of the page
+     *
+     * @param {Object} mainContainer - the main page container (mid-level)
```
What does "mid-level" and "top-level" mean?

BlueSlug replied:
I've removed any logic relating to mainContainer, so the distinction between the elements' hierarchy is no longer relevant

cindyli commented on src/ui/dynamicViewComponentManager.js:
```diff
@@ -89,11 +89,13 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
-     * - "managedComponent": the new view component to register
-     * - "completionEvent": the event to be fired upon successful completion
+     *
+     * @param {Object} that - the dynamicViewComponentManager itself
```
{Object} -> {Component} to line 96, 97

cindyli commented on src/ui/dynamicViewComponentManager.js:
```diff
@@ -103,11 +105,13 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
-     * - "managedComponentIndividualClass": the CSS control class of the view component
-     * - "completionEvent": the event to be fired upon successful completion
+     *
+     * @param {Object} that - the dynamicViewComponentManager itself
```
Same as above and other places.

cindyli commented on src/ui/storyTellingServerUI.js:
```diff
@@ -26,12 +30,16 @@ sjrk.storyTelling.getParameterByName = function (name, url) {
+/**
+ * Loads a story View page and a particular story from a story ID from the query string
+ *
+ * @param {Object} clientConfig - a collection of client config values consisting of
```
Another place that @typedef should be considered. An example is [here](https://github.com/GPII/universal/blob/master/gpii/node_modules/flowManager/src/PSPChannel.js#L177-L184).

cindyli commented on src/ui/storyTellingServerUI.js:
```diff
@@ -76,12 +84,16 @@ sjrk.storyTelling.loadStoryFromParameter = function (clientConfig, options) {
+/**
+ * Loads a story Browse page and populates it with a set of stories
+ *
+ * @param {Object} clientConfig - a collection of client config values consisting of
```
Same as above here and at line 152.

cindyli commented on src/ui/templateManager.js:
```diff
@@ -134,42 +134,49 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+    /**
+     * Injects the provided template content to the partial templates collection of the template renderer
+     *
+     * @param {Object} templateRenderer - the template renderer
```
{Object} -> {Component}

cindyli commented on src/ui/transforms.js:
```diff
@@ -15,17 +15,23 @@ https://raw.githubusercontent.com/fluid-project/sjrk-story-telling/master/LICENS
+    * @param {String} transformSpec.[delimiter] - the delimiter of terms within the given strings, defaults to ","
+    * @param {Boolean} transformSpec.[trim] - flag to trim excess whitespace from each term. defaults to true
+    *
+    * @return {String[]} - the resulting array
```
{String[]} -> {Array}. Also, please expand on "the resulting array".

cindyli replied:
The comment above applies across this pull request.

BlueSlug committed:
SJRK-321: removed extra args & linter exceptions from db validation func

BlueSlug committed:
SJRK-321: JSDoc tidying, added a type definition for ClientConfig object

BlueSlug committed:
SJRK-321: updating JSDoc types and function documentation clarificaiton

BlueSlug committed:
SJRK-321: blockContentValues -> blockFields, and related docs updates

BlueSlug committed:
SJRK-321: setAuthoringEnabledClass -> showEditPageContainer, refactored

BlueSlug committed:
SJRK-321: refining JSDoc types and some light refactoring

BlueSlug committed:
SJRK-321: more JSDoc clarification

BlueSlug commented:
Ready for more review

cindyli commented:
This pull request looks good. I will merge it once the corresponding CI deployment pull request https://github.com/inclusive-design/ci-service/pull/37 is merged.

@BlueSlug, could you coordinate with @waharnum regarding the merge of the ci-service pull? Thanks.

BlueSlug commented:
@cindyli the CI pull request has been merged: https://github.com/inclusive-design/ci-service/pull/37

thanks @waharnum!

cindyli commented:
Merged at 87aa0b3471c74b002147e93d5e0219efaabf4546
